### PR TITLE
minimal decks with maximized Library

### PR DESCRIPTION
**Pull request by ronso0**
picking up https://bugs.launchpad.net/mixxx/+bug/1467004, #878

Let's discuss what's needed for those decks.
They should facilitate picking a matching next track, so for now built in the basic track infos only:
- Titel / Artist
- BPM / Key
- track time (it would be nice to force 'remaining' only, and also hide the centiseconds)
- track overview
- Spinny / cover
- Play indicator

**The use cases I have in mind:**
**1** Live set, prepare next tracks
Have an eye on the running track(s) while diving into your library.
Compare key and BPM for next tracks.

**2** minimal decks for controller users
...and/or those who don't need the big cover/spinny, mixer, effects indicators.
All hotcues are visible and accessible via controlller LEDs and buttons.
Have a large library even with 4 decks.
Deck controls (quantize, keylock,...) and effects can be adjusted with the regular view.

**3** minimal decks for AutoDJ

**4** ?

~~I used the existing fixed-size templates for Spinny/cover for now. If we make them smaller and replace the track overview with a simple progress bar the decks could shrink even more.~~
I'll try to make the decks all 40px tall. Spinny won't be of any use then IMO but at least its appearance doesn't change when maximizing the library, and it still serves as a PLay indiicator.

minimal-tango

minimal-shade

minimal-deere

minimal-late

### CODE REVIEW

**Comment by daschuer** [comment]
Great! For me this is already in a mergeable state.
I have never noticed before that we have just "by" if no track is loaded. You can use the "track_samples" CO and hide if it is 0.0.

I think there is no different use case for the time display. I would expected that that the display is not changed when expand the library. 

I think the minimal deck can be more mini. 
What should be done with the mini decks? We can't currently cue the track. Because no cue Headphones and xfader access. What is the use for the play button?

Currently it serves perfectly the "compare properties use case".

**Comment by ronso0** [comment]
This is still just a draft and should not be merged, yet.

I updated the 1st post with the use cases.

The decks' height is now determined by the Cover art/Spinny size. If that was even smaller I guess it'd be hardly recognisable.

> I have never noticed before that we have just "by" if no track is loaded. You can use the "track_samples" CO and hide if it is 0.0.

yeah, or with `...,track_loaded`.
Would be nice if that string could be translated.

**Comment by daschuer** [comment]
A straight forward solution for translation is to introduce WBy. This way the text is in the c++ domain.

**Comment by daschuer** [comment]
Or WLabelTranslated with a lookup table of various useful strings.

**Comment by nopeppermint** [comment]
I like the idea of those minimal decks

> the basic track infos only:
> - Titel / Artist
> - BPM / Key
> - track time (it would be nice to force 'remaining' only, and also hide the centiseconds)
> - track overview
> - Spinny / cover
> - Play indicator

is a good idea

**Comment by ronso0** [comment]
@daschuer While testing how it could like in Shade, I notice Shade doesn't show the track duration...on purpose?

**Comment by daschuer** [comment]
> @daschuer While testing how it could like in Shade, I notice Shade doesn't show the track duration...
on purpose?

In the library or decks? I can see it.

**Comment by ronso0** [comment]
okay, only Tango & LateNight have the track duration displayed in the decks. Never noticed before..

**Comment by ronso0** [comment]
@daschuer Do you agree with this layout for Shade?
It's just a mockup, i.e. the Play button can be optimized.
shade-mini

**Comment by daschuer** [comment]
Great :-) Thank you very much.

**Comment by nopeppermint** [comment]
nice!

maybe it makes sense to display this minimal deck also with not maximized library, could then be enabled as feature on the skin-settings menu?

**Comment by ronso0** [comment]
yeah, this is also one of the use cases I listed above.
but for now let's show them only above the maximized library. I'm not sure if all Shade users like it that way, maybe we need a menu item for that

**Comment by ronso0** [comment]
After finishing the implementation for Shade and looking at LateNight right now, I think it makes sense to use the same (Shade) layout for all skins. 

Only difference between the skins would be the a/symmetry (Tango) and the optional Cover/Spinny (all except Shade).

**Comment by ronso0** [comment]
LateNight at minimal size:
minimal-latenight

**Comment by ronso0** [comment]
-updated screens in first post-

**Comment by daschuer** [comment]
Thank you for the nice feature. I think I will test it on my old netbook before merge. 
Do you expect any performance regression? Do we have two waveform widget per deck?

**Comment by ronso0** [comment]
I can't tell if it really affects performance, but yes, there are now 2 overviews per deck.
If you notice issues on your netbook, I could load them as singletons, but I'm not sure if the resizing works then

**Comment by daschuer** [comment]
Currently building, takes usually ages. Is the singleton solution a big deal? Do you know if the waveform code is executed twice? I am curious to see the results.

**Comment by ronso0** [comment]
I can do a simple CPU comparison with htop later on.
Implementing singletons is no big deal itself, but IIRC it requires nesting and some hacks sometimes. For example the (scrolling) waveform solution wouldn't work the same in LateNight and Deere..and the spinnies in Deere force-(re)open the waveform splitter.
So let's see what htop shows.

> Currently building, takes usually ages

why bother building this PR on the netbook? it's just skin changes

**Comment by daschuer** [comment]
I want to try the current master there anyway.

**Comment by nopeppermint** [comment]
I like to test this but current master is not building on my Linux, does someone else has similar errors?
https://mixxx.zulipchat.com/#narrow/stream/109122-general/subject/master_no_longer_builds/near/152352918

**Comment by ronso0** [comment]
@nopeppermint take a build from http://downloads.mixxx.org/builds/master/release/ ?

**Comment by nopeppermint** [comment]
@ronso0 
thanks for the tip, i installed a .deb version and copied the skin folder

I just tested the latenight version and there the gui is hanging with the follwing error when I switch to biglibraryies:
```
Debug [Main]: Committing transaction on "MIXXX-1" result: true
Warning [Main]: QXcbConnection: XCB error: 8 (BadMatch), sequence: 33563, resource id: 96469255, major code: 155 (Unknown), minor code: 11
Debug [Main]: OUT OF CONTROL 4 427336 0 5 4
Debug [Main]: OUT OF CONTROL 4 427336 0 6 5
Debug [Main]: OUT OF CONTROL 4 427336 0 7 6
Debug [Main]: OUT OF CONTROL 4 427336 0 8 7
Debug [Main]: OUT OF CONTROL 4 427336 0 9 8
Debug [Main]: OUT OF CONTROL 4 427336 0 10 9
```

all the other skins are working

**Comment by ronso0** [comment]
@nopeppermint 
I have no idea what tha message means..
With the last commit I unified the waveform hack, that also eliminates qpainter log messages.
Please test again!

**Comment by ronso0** [comment]
@daschuer 
CPU load is the same with minimal decks (no spinnies, no waveforms).
Memory usage is a little higher: 726M (master) vs 770M (minimal-decks)

ToDo:
* Coverart/Spinny Singletons in all skins
* Overview Singletons in all skins

**Comment by nopeppermint** [comment]
@ronso0  now it's working

**Comment by ronso0** [comment]
> Memory usage is a little higher: 726M (master) vs 770M (minimal-decks)

..and goes down to 620M with Singletons for spinny and overview

**Comment by ronso0** [comment]
> With the expanded library view the CPU load drops significant. I assume because of the missing meters.

It drops here as well and I have hidden the mixer, I guess it's the missing waveforms. #1897 ?

> there is a significant difference in CPU load in expanded library view between enabled and disabled waveform.

what do you mean by 'enabled'?

**Comment by ronso0** [comment]
> Grate. :-) LGTM Thank you very much. Is this still ready for merge?

On my machine the singletons reduced the memory load by 100M, so I'd rather have another look at the other skins and improve it there, as well.
And I'd love to have this tested more before merge, to find bugs like @nopeppermint did.

**Comment by nopeppermint** [comment]
looks good!

=> I suggest to add at least the "eject Button" 

=> maybe grey out the skin settings which no longer are active if in "big library" mode, as in Latenight/Tango/Deere the skin settings button is always visible.

Settings with no function on "big library":
Latenight:
- 8 Hotcue
- Rate Control
- Vinyl Control
- Big Spinny/Cover Art
- all Mixer, Effects and Samplers Settings

Deere:
- Parallel Waveforms
- Spinnies
- bpm
- Star Rating
- 8 Hotcues
- Vinyl Control Options
- all Mixer, Samplers and Effects Setttings

Tango:
- Symmetric Overviews
- Symmetric Time/Duration
- Hot Cues
- Loop / Beatjump Buttons
- Permanent Deck Controls
- BPM/Sync & Pitch Fader options
- Track Comment
- Cover Art Size
- Channel Mixer options
- Always Show Level Meters
- Effects Options
- Sampler options

File: res/skins/Shade/skin.xml [hunk 1]
```diff
@@ -427,14 +454,36 @@
...
         </Template>
+
+        <!-- minimal decks, visible with maximized library -->
+        <Template src="skin:decks_row_small.xml"/>
+
+        <!-- put invisble waveforms Singletons here to enable end-of-track warning
+            without actually showing scrolling waveforms -->
+        <WidgetGroup>
+          <Layout>horizontal</Layout>
+          <Size>0me,0f</Size>
```
Comment (daschuer) [inline res/skins/Shade/skin.xml]: This does not work. The waveform width is reported as 5xx and visible. The width changes from 3xx when switching library though.
So it is tendered still rendered.

Do you have an idea why this is required for the end of Treck Warnung? We need probably an alternative solution.

**Review by daschuer — CHANGES_REQUESTED** [review]
I think we have the choice to remove the end of track solution, or just except the higher cpu. It has to be available anyway for normal view.

File: res/skins/Shade/skin.xml (hunk 1 above)
Comment (ronso0) [reply res/skins/Shade/skin.xml]: yeah, as stated above this is just a hack (1) for the end-of-track warning.
and I recall a bug (2) where needed these invisible waveforms to keep the rest of the GUI responsove with Intel graphics when waveforms are hidden in Tango (which we now apply to all skins to have only minimal decks & library).
https://bugs.launchpad.net/mixxx/+bug/1758660
https://bugs.launchpad.net/mixxx/+bug/1204294
https://bugs.launchpad.net/mixxx/+bug/1801636

If someone could look into the issue, that would be very nice as we could remove some skin hacks if it's fixed.

Comment (ronso0) [reply res/skins/Shade/skin.xml]: >  The waveform width is reported as 5xx and visible.

can you actually see it in Shade? or is debug telling the widget is visible?

**Comment by daschuer** [comment]
here is a fix for end_of_track https://github.com/mixxxdj/mixxx/pull/1963

**Comment by ronso0** [comment]
> I suggest to add at least the "eject Button"

yeah. where should we put it where it's not pressed by accidence?

**Comment by ronso0** [comment]
> => maybe grey out the skin settings which no longer are active if in "big library" mode, as in Latenight/Tango/Deere the skin settings button is always visible.
>
> Settings with no function on "big library":
Latenight: ...

Hmm although you're right I doubt there'll be such a confusion that requires making the skin settings more complex (either nesting buttons in the menu, or creating even more different types of menu button templates which include the nesting and conditional visibility)
AFAICT the only settings that might be used with maximized library are:
- show 4 decks
- show Preview deck
- show Cover Art in library
- (maybe Spinny settings, but actually I'd expect them to persist when switching to max. libary)

I guess it's easier to create tool bar buttons for those toggles.

**Comment by nopeppermint** [comment]
> I guess it's easier to create tool bar buttons for those toggles.

But even with not displayed skin settings button you could open the skin settings over shortcut "CTRL+1" and the menu button "View/Show Skin Settings Menu"

here my PR: https://github.com/ronso0/mixxx/pull/4

Skin Settings Button not visible with big library, additional toolbar settings with big library

normal_library
big_library

**Comment by nopeppermint** [comment]
@ronso0 what do you think?

I would modify the other skins as well.
I suggest to not  display the skin settings knob an add those toolbar buttons;

deere:
- 4 deck
- cover art
- artist & track title (maybe show title always with big library)
- preview deck
- cover art in library

tango:
- 4 deck
- spinny
- cover art
- preview deck
- cover art in library

**Comment by nopeppermint** [comment]
I just found out that the length of the waveform in the small decks depend on the text length of the title, what do you think about a fixed waveform length with big library?
width_of_small_deck_waveform

**Comment by nopeppermint** [comment]
I updated my PR: ronso0#4
I now also added a split cue button and knobs to big library toolbar:
- head mix
- head gain
- balance
- master gain
knobs_smal

**Comment by nopeppermint** [comment]
I just found  out that "Loop/Beatjump Buttons" are no longer displayed with this PR on Skin Tango/Tango64 the skin-settings does not change their visibility, they are always NOT visible

**Comment by Be-ing** [comment]
That bug was just fixed in #1964, so merging master should fix it.

**Comment by ronso0** [comment]
> > I suggest to add at least the "eject Button"
> > yeah. where should we put it where it's not pressed by accidence?
> 
> between bpm and key, as those two fields are not clickable, but play and playpositiontext are.
> [ronso0@9542200](https://github.com/ronso0/mixxx/commit/9542200ff0d113c56cfd1643daa49d34c8fecab8)
> 
> eject

Thanks for the example. For me, it feels a bit unpleasant to separate Key/BPM, maybe Eject can go to the far left next to the overview? or better in between time and key display as clicking the time doesn't affect playback?

**Comment by ronso0** [comment]
> I updated my PR: [ronso0#4](https://github.com/ronso0/mixxx/pull/4)
> I now also added a split cue button and knobs to big library toolbar:
>     * head mix
>     * head gain
>     * balance
>     * master gain

For me, the reason to work on minimal decks is to have a clean, minimal interface while searching for tracks. That's why I think we can omit the master/headphone mixer with maximized library.
I someone needs them he/she can hit the spacebar and switch to regular full-featured view easily.

What do others think?

**Comment by daschuer** [comment]
Yes that is my use case as well, searching for a new track while having an eye of the deck. 
I can imagine that I one want to use the headphone Mixxx button in this case, to monitor master, especially if I hear only unusable bassy garbage from the main PA.
This happens if you preview a track.

This does not effect me though because I have the knob on my controller. My favorite knob :-).

**Comment by nopeppermint** [comment]
> maybe Eject can go to the far left next to the overview? or better in between time and key display as clicking the time doesn't affect playback?

where you want, I don't care so much where eject is placed

**Comment by nopeppermint** [comment]
@ronso0  what do you think about https://github.com/ronso0/mixxx/pull/4/commits/cbbfe034b17ca0aca64e57af803c355b30e67dee (independent of the requirements of knobs in the toolbar)
grafik

**Comment by nopeppermint** [comment]
https://github.com/ronso0/mixxx/pull/4/commits/259491eed9705dda2f12351eb4f108b512238a75
small deck as skin setting, additional knobs only visible when small deck selected in skin settings
smal_deck_skin_setting

**Comment by Be-ing** [comment]
>  small deck as skin setting, additional knobs only visible when small deck selected in skin settings

I think including anything other than the decks and library in this view kinda defeats the purpose.

**Comment by ronso0** [comment]
> I think including anything other than the decks and library in this view kinda defeats the purpose.

Yep! Otoh "minimal decks" as skin options is a valid preference, but -yes- it goes way beyond of wwhat I had in mind with this PR.
In general, and whatver possibilities we have at our hands with the (to develop) QML option [layout] + [color scheme], I think it's a mssive overload to introduce both 'minimal decks' with max. library AND reular view with even more options.
Considering my fellow DJs struggling with Serato and Recordbox and alike, I'd say we have more than enough optioins available already to please quite a lot of habits...more than enough to make it even more complex to choose, no to mention the controls to mak'em (easily) controllable with MIID controllers

**Comment by Be-ing** [comment]
Considering the small amount of space, I think I'd prefer to not have the eject button in these mini decks and have a little more width for title/artist name.

**Comment by ronso0** [comment]
> The only information I miss from the mini decks is the artist name. Maybe move the time to the top right and put the artist where the time is now?

I liked it that the time is close to the overview, but I'll try it.

I can try to put Eject into a horizontal group with the overview, so artist/time/etc can still take half of the available h-space.

**Comment by ronso0** [comment]
I added Artist, but with current master I experience annoying repaint issues (time display gets cropped) when I cycle through elapsed/remaining/both. I don't know what to do about it, no Size/SizePolicy possibilities can solve this afaict..
Can anyone confirm this behaviour?

**Comment by ronso0** [comment]
[ToDo]
- make overviews Singletons in all skins
- check if #1963 allows to remove invisible waveforms with max library

**Comment by ronso0** [comment]
> I added Artist, but with current master I experience annoying repaint issues (time display gets cropped) when I cycle through elapsed/remaining/both.

...with very long track titles.

**Comment by ronso0** [comment]
Finished, please test!

also, please test with long track titles and check if ugly things happen (overlaps, cropped time display etc). overviews are singletons now, please check if this works reliably on your OSs.

**Comment by ronso0** [comment]
If all works well, let's merge this and discuss skin settings and tool bar options elsewhere.

**Comment by ronso0** [comment]
[ToDo]
* check if #1963 allows to remove invisible waveforms with max library

will have to build it for the affected machine, could take a few days to get my hands on it again..

**Comment by daschuer** [comment]
https://github.com/mixxxdj/mixxx/pull/1963 is merged. So we can remove the end of track workaround here.

**Comment by ronso0** [comment]
I removed invisble waveforms and tested master (with #1963) on a previously affected machine.
GUI remains agile, no regression. Thanks @daschuer for fixing that!

Ready to merge as soon is this is positively tested on other OS tha linux
> also, please test with **long track titles** with minimal window size and check if ugly things happen (overlaps, cropped time display etc). overviews are singletons now, please check if this works reliably on your OSs.

**Comment by nopeppermint** [comment]
I just rebuild this on Linux Mint 19.1 x64 and Latenight/Tango/Tango64/Deere still doesn't show's the waveforms..

Or to be more precise on startup they are there and then after changing skin to latenight they are gone on all skins..

But when I start with Latenight the waveforms are visible, but as soon as I switch skin they are gone.
**So it seems like switching between skins cause no longer displaying the waveforms !!**

normal_lib
big_lib

**Comment by ronso0** [comment]
I experienced this a few times, but only in dev mode where some things get screwed up after reloading skins many times. I'll have a look.
@daschuer might this happen if overview Singletons have the same name in different skins? Are they cleared when unloading a skin?

**Comment by daschuer** [comment]
I can confirm the missing waveform. But it is only the actual waveform data that is missing, the decoration including end of track signal works.

**Comment by ronso0** [comment]
> So it seems like switching between skins cause no longer displaying the waveforms (overview)

I can confirm this with master. Samplers are affected as well, so it apparently it has nothing to do with Singletons

**Comment by daschuer** [comment]
https://github.com/mixxxdj/mixxx/pull/1986 fixes the remaining issue.

**Comment by nopeppermint** [comment]
> #1986 fixes the remaining issue.

indeed, nice! 
tested on Linux Mint 19.1 x64 QT 5.9.5

**Comment by ronso0** [comment]
This ir ready to merge. Anyone had the time to test?

**Comment by ronso0** [comment]
Ready to roll!

**Comment by Be-ing** [comment]
Nice work, thank you! Sorry for the delay merging this.

I noticed when testing that when I switch skins in the preferences while the library is maximized, when the new skin is loaded, the library is not maximized. Could you follow this up with another PR to use the same CO for maximizing the library in each skin?

**Comment by ronso0** [comment]
We use `[Master],maximize_library` in all skins, and I think we decided to load each skin with default view to avoid confusion "[Master],maximize_library","0"
Is it really desired to change this? IIRC for example in Shade there's no GUI switch to toggle max library

**Comment by Be-ing** [comment]
Oh, that makes sense. Switching skins while having the library maximized is a very niche use case, so I agree that it is better to leave it as is with setting `[Master], maximize_library` to 0 on skin load.

**Comment by ronso0** [comment]
okk. Thanks!

**Pull request merged**